File: src/fppy/io/legacy_data.py
"""Parsers for legacy FAIR template data blocks."""

from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path

import pandas as pd

from fppy.config import default_legacy_config

_NUMBER_RE = re.compile(r"[+-]?(?:\d+(?:\.\d*)?|\.\d+)(?:[Ee][+-]?\d+)?")
_WORD_RE = re.compile(r"[^\s;]+")


@dataclass(frozen=True)
class LegacyTemplateBundle:
    """Parsed legacy template artifacts loaded from default paths."""

    fmdata: pd.DataFrame
    fmage: pd.DataFrame
    fmexog: pd.DataFrame


def generate_smpl_period_index(start: str | float | int, end: str | float | int) -> pd.Index:
    """Generate a quarter-period index from an inclusive SMPL window."""

    start_year, start_q = _parse_smpl_token(str(start))
    end_year, end_q = _parse_smpl_token(str(end))
    start_code = _smpl_code(start_year, start_q)
    end_code = _smpl_code(end_year, end_q)

    if end_code < start_code:
        raise ValueError("SMPL end period must not come before start period.")

    periods: list[str] = []
    for code in range(start_code, end_code + 1):
        year, quarter = _smpl_from_code(code)
        periods.append(f"{year}.{quarter}")

    return pd.Index(periods, name="smpl")


def parse_fmdata_text(text: str) -> pd.DataFrame:
    """Parse legacy `fmdata` text into a wide pandas frame."""

    return _parse_fm_numeric_file(text, block_name="fmdata")


def parse_fmdata_file(path: Path | str) -> pd.DataFrame:
    """Parse an `fmdata` file from disk."""

    text = _read_text(path)
    return parse_fmdata_text(text)


def parse_fmage_text(text: str) -> pd.DataFrame:
    """Parse legacy `fmage` text into a wide pandas frame."""

    return _parse_fm_numeric_file(text, block_name="fmage")


def parse_fmage_file(path: Path | str) -> pd.DataFrame:
    """Parse an `fmage` file from disk."""

    text = _read_text(path)
    return parse_fmage_text(text)


def parse_fmexog_text(text: str) -> pd.DataFrame:
    """Parse legacy `fmexog` CHANGEVAR instructions into a structured frame."""

    rows: list[dict[str, object]] = []
    lines = text.splitlines()

    current_window_start: str | None = None
    current_window_end: str | None = None
    current_window_size: int | None = None
    in_changevar = False

    pending_scalar = False
    pending_variable: str | None = None
    pending_method: str | None = None
    pending_values: list[float] = []
    pending_vector = False

    def _close_pending() -> None:
        nonlocal \
            pending_scalar, \
            pending_variable, \
            pending_method, \
            pending_values, \
            pending_vector, \
            rows
        if pending_variable is None:
            return
        if current_window_start is None or current_window_end is None:
            raise ValueError("CHANGEVAR instruction found outside an SMPL window.")
        if not pending_values:
            raise ValueError(f"No values parsed for instruction {pending_variable}.")
        rows.append({
            "window_start": current_window_start,
            "window_end": current_window_end,
            "variable": pending_variable,
            "method": pending_method,
            "is_vector": bool(pending_vector),
            "n_values": len(pending_values),
            "values": tuple(pending_values),
        })
        pending_scalar = False
        pending_variable = None
        pending_method = None
        pending_values = []
        pending_vector = False

    for raw_line in lines:
        line = raw_line.strip()
        if not line:
            if pending_vector:
                continue
            if pending_scalar:
                continue
            continue
        if line.startswith("@"):  # comment marker in legacy files
            continue

        line_has_semicolon = ";" in line
        tokens = _tokens(line)
        if not tokens:
            if (pending_scalar or pending_vector) and line_has_semicolon:
                _close_pending()
            continue

        first = tokens[0].upper()
        if first == "SMPL":
            if len(tokens) < 3:
                raise ValueError("SMPL statements require start and end periods.")
            _close_pending()
            current_window_start = _normalize_period_token(tokens[1])
            current_window_end = _normalize_period_token(tokens[2])
            # Require valid window ordering.
            current_window_size = len(
                generate_smpl_period_index(current_window_start, current_window_end)
            )
            in_changevar = False
            continue

        if first == "CHANGEVAR":
            if current_window_start is None:
                raise ValueError("CHANGEVAR encountered before SMPL statement.")
            _close_pending()
            in_changevar = True
            continue

        if first == "RETURN":
            _close_pending()
            break

        if not in_changevar:
            continue

        if pending_scalar or pending_vector:
            # CHANGEVAR instructions can list values on subsequent lines. In
            # modern FAIR scenario decks we also see vector blocks that omit a
            # ';' on the header line (for example `JGPHASE ADDDIFABS` followed
            # by one numeric value per period). Treat contiguous numeric-only
            # lines as values for the active instruction and promote scalar
            # instructions to vector when multiple values are observed.
            if not _is_numeric_token(tokens[0]):
                # Next instruction header encountered.
                _close_pending()
                # Fall through to parse this line as a new instruction header.
            else:
                values = _extract_numbers(line)
                if values:
                    pending_values.extend(values)
                    if pending_scalar and len(pending_values) > 1:
                        pending_scalar = False
                        pending_vector = True

                if (
                    pending_vector
                    and current_window_size is not None
                    and len(pending_values) >= current_window_size
                ):
                    if len(pending_values) > current_window_size:
                        raise ValueError(
                            "Too many values parsed for vector instruction "
                            f"{pending_variable}: expected {current_window_size}, found {len(pending_values)}."
                        )
                    _close_pending()
                    continue

                if line_has_semicolon:
                    _close_pending()
                continue

        # A new instruction header.
        if _is_legacy_end_marker(line):
            continue

        variable = tokens[0]
        method: str | None = None
        is_vector = False
        value_text = line

        if len(tokens) >= 2:
            # Method present only when the second token is non-numeric and non-delimiter.
            if not _is_numeric_token(tokens[1]):
                method = tokens[1].strip(";'\"")
                value_text = line[line.find(method) + len(method) :]
                if line_has_semicolon:
                    is_vector = True

        if not line_has_semicolon and len(tokens) == 1:
            # Scalar with method omitted or with explicit scalar value on this line.
            is_vector = False
        elif line_has_semicolon and len(tokens) == 1:
            # VAR ; vector with no method
            is_vector = True

        extracted = _extract_numbers(value_text)

        if is_vector:
            pending_scalar = False
            pending_vector = True
            pending_variable = variable
            pending_method = method
            pending_values = list(extracted)
            if pending_values and line_has_semicolon:
                _close_pending()
            continue

        pending_scalar = True
        pending_vector = False
        pending_variable = variable
        pending_method = method
        pending_values = list(extracted)

        if pending_values:
            _close_pending()

    _close_pending()

    return pd.DataFrame(
        rows,
        columns=[
            "window_start",
            "window_end",
            "variable",
            "method",
            "is_vector",
            "n_values",
            "values",
        ],
    )


def parse_fmexog_file(path: Path | str) -> pd.DataFrame:
    """Parse a legacy `fmexog` file from disk."""

    text = _read_text(path)
    return parse_fmexog_text(text)


def load_default_legacy_templates() -> LegacyTemplateBundle:
    """Load `fmdata`, `fmage`, and `fmexog` from default template paths."""

    paths = default_legacy_config().legacy
    return LegacyTemplateBundle(
        fmdata=parse_fmdata_file(paths.fmdata),
        fmage=parse_fmage_file(paths.fmage),
        fmexog=parse_fmexog_file(paths.fmexog),
    )


def _parse_fm_numeric_file(text: str, *, block_name: str) -> pd.DataFrame:
    lines = text.splitlines()
    current_periods: pd.Index | None = None
    series_by_variable: dict[str, pd.Series] = {}

    i = 0
    while i < len(lines):
        raw_line = lines[i].strip()
        i += 1

        if not raw_line or raw_line.startswith("@"):
            continue

        tokens = _tokens(raw_line)
        if not tokens:
            continue

        command = tokens[0].upper()
        if command == "SMPL":
            if len(tokens) < 3:
                raise ValueError("SMPL statement missing boundaries.")
            current_periods = generate_smpl_period_index(tokens[1], tokens[2])
            continue

        if command != "LOAD":
            continue

        if current_periods is None:
            raise ValueError(f"LOAD block for {block_name} before SMPL statement.")
        if len(tokens) < 2:
            raise ValueError(f"LOAD block in {block_name} missing variable name.")

        variable = tokens[1]
        values: list[float] = []
        closed = False

        while i < len(lines):
            current = lines[i].strip()
            if not current:
                i += 1
                continue
            if _is_end_marker(current):
                closed = True
                i += 1
                break

            numbers = _extract_numbers(current)
            if not numbers:
                raise ValueError(f"Expected numeric values in {block_name} LOAD {variable}.")
            values.extend(numbers)
            i += 1

        if not closed:
            raise ValueError(
                f"Unterminated LOAD block for {variable} in {block_name} block "
                f"{current_periods[0]}..{current_periods[-1]}."
            )

        if len(values) != len(current_periods):
            raise ValueError(
                f"Value count mismatch for {block_name} variable {variable}: "
                f"expected {len(current_periods)} values, found {len(values)}."
            )

        _set_block_values(series_by_variable, variable, current_periods, values)

    if not series_by_variable:
        return pd.DataFrame()

    frame = pd.concat(series_by_variable, axis=1).sort_index()
    frame.index.name = "smpl"
    return frame


def _set_block_values(
    series_by_variable: dict[str, pd.Series],
    variable: str,
    periods: pd.Index,
    values: list[float],
) -> None:
    block = pd.Series(values, index=periods, name=variable, dtype="float64")
    existing = series_by_variable.get(variable)

    if existing is None:
        series_by_variable[variable] = block
        return

    all_index = existing.index.union(periods)
    updated = existing.reindex(all_index)
    updated.loc[periods] = block
    series_by_variable[variable] = updated.astype("float64")


def _read_text(path: Path | str) -> str:
    return Path(path).read_text(encoding="utf-8", errors="replace")


def _tokens(line: str) -> list[str]:
    return [token.strip("'\"") for token in _WORD_RE.findall(line)]


def _extract_numbers(line: str) -> list[float]:
    return [float(match.group(0)) for match in _NUMBER_RE.finditer(line)]


def _parse_smpl_token(token: str) -> tuple[int, int]:
    cleaned = token.strip().strip("'\"").strip()
    if "." not in cleaned:
        raise ValueError(f"Invalid SMPL token {token!r}; expected <year>.<quarter>.")

    year_text, quarter_text = cleaned.split(".", 1)
    if not year_text or not quarter_text:
        raise ValueError(f"Invalid SMPL token {token!r}; expected <year>.<quarter>.")

    if not year_text.lstrip("+").replace("-", "", 1).isdigit():
        raise ValueError(f"Invalid SMPL year token {token!r}.")

    if not quarter_text.isdigit():
        raise ValueError(f"Invalid SMPL quarter token {token!r}.")

    year = int(year_text)
    quarter = int(quarter_text)

    if quarter not in (1, 2, 3, 4):
        raise ValueError(f"SMPL quarter must be in 1..4; got {quarter} from {token!r}.")

    return year, quarter


def _normalize_period_token(token: str) -> str:
    year, quarter = _parse_smpl_token(token)
    return f"{year}.{quarter}"


def _smpl_code(year: int, quarter: int) -> int:
    return year * 4 + quarter - 1


def _smpl_from_code(code: int) -> tuple[int, int]:
    year, remainder = divmod(code, 4)
    return year, remainder + 1


def _is_legacy_end_marker(line: str) -> bool:
    return bool(re.fullmatch(r"[^A-Za-z0-9]*END[^A-Za-z0-9]*", line.strip(), re.IGNORECASE))


def _is_end_marker(line: str) -> bool:
    return bool(re.fullmatch(r"[^A-Za-z0-9]*END[^A-Za-z0-9]*", line.strip(), re.IGNORECASE))


def _is_numeric_token(token: str) -> bool:
    return bool(_NUMBER_RE.fullmatch(token.strip("'\"")))

File: fp-r/R/legacy_data.R
generate_smpl_periods <- function(start, end) {
  seq_periods(as.character(start), as.character(end))
}

parse_legacy_numeric_tokens <- function(line) {
  matches <- gregexpr(
    "(?:[+-]?(?:\\d+(?:\\.\\d*)?|\\.\\d+)(?:[Ee][+-]?\\d+)?|NA|NAN)",
    line,
    perl = TRUE,
    ignore.case = TRUE
  )
  found <- regmatches(line, matches)[[1]]
  if (!length(found)) {
    return(numeric())
  }
  vapply(found, function(token) {
    upper <- toupper(trimws(token))
    if (upper %in% c("NA", "NAN")) {
      return(NA_real_)
    }
    as.numeric(token)
  }, numeric(1))
}

parse_fm_numeric_text <- function(text, block_name = "fmdata") {
  lines <- strsplit(gsub("\r", "", text), "\n", fixed = TRUE)[[1]]
  current_periods <- NULL
  series_by_variable <- list()
  i <- 1L

  while (i <= length(lines)) {
    raw_line <- trimws(lines[[i]])
    i <- i + 1L

    if (!nzchar(raw_line) || grepl("^\\s*@", raw_line)) {
      next
    }

    tokens <- strsplit(raw_line, "\\s+", perl = TRUE)[[1]]
    if (!length(tokens)) {
      next
    }

    command <- toupper(tokens[[1]])
    if (identical(command, "SMPL")) {
      if (length(tokens) < 3L) {
        stopf("SMPL statement missing boundaries in %s", block_name)
      }
      current_periods <- generate_smpl_periods(tokens[[2]], tokens[[3]])
      next
    }

    if (!identical(command, "LOAD")) {
      next
    }
    if (is.null(current_periods)) {
      stopf("LOAD block for %s before SMPL statement", block_name)
    }
    if (length(tokens) < 2L) {
      stopf("LOAD block in %s missing variable name", block_name)
    }

    variable <- gsub("^['\"]|['\"]$", "", tokens[[2]])
    variable <- sub(";\\s*$", "", variable)
    values <- numeric()
    closed <- FALSE

    while (i <= length(lines)) {
      current <- trimws(lines[[i]])
      i <- i + 1L

      if (!nzchar(current)) {
        next
      }
      if (grepl("^\\s*'END'\\s*$", current, perl = TRUE) || grepl("^\\s*END\\s*;?\\s*$", current, perl = TRUE)) {
        closed <- TRUE
        break
      }
      numbers <- parse_legacy_numeric_tokens(current)
      if (!length(numbers)) {
        stopf("Expected numeric values in %s LOAD %s", block_name, variable)
      }
      values <- c(values, as.numeric(numbers))
    }

    if (!closed) {
      stopf(
        "Unterminated LOAD block for %s in %s block %s..%s",
        variable,
        block_name,
        current_periods[[1]],
        current_periods[[length(current_periods)]]
      )
    }
    if (length(values) != length(current_periods)) {
      stopf(
        "Value count mismatch for %s variable %s: expected %d, found %d",
        block_name,
        variable,
        length(current_periods),
        length(values)
      )
    }

    existing <- series_by_variable[[variable]]
    block <- setNames(as.numeric(values), current_periods)
    if (is.null(existing)) {
      series_by_variable[[variable]] <- block
      next
    }
    combined_names <- unique(c(names(existing), names(block)))
    combined <- rep(NA_real_, length(combined_names))
    names(combined) <- combined_names
    combined[names(existing)] <- as.numeric(existing)
    combined[names(block)] <- as.numeric(block)
    series_by_variable[[variable]] <- combined
  }

  if (!length(series_by_variable)) {
    return(data.frame(stringsAsFactors = FALSE, check.names = FALSE))
  }

  all_periods <- sort(unique(unlist(lapply(series_by_variable, names))), decreasing = FALSE)
  frame <- data.frame(
    period = all_periods,
    stringsAsFactors = FALSE,
    check.names = FALSE
  )
  for (name in names(series_by_variable)) {
    values <- rep(NA_real_, length(all_periods))
    names(values) <- all_periods
    values[names(series_by_variable[[name]])] <- as.numeric(series_by_variable[[name]])
    frame[[name]] <- as.numeric(values)
  }
  frame
}

parse_fm_numeric_file <- function(path, block_name = NULL) {
  normalized_path <- normalizePath(path, winslash = "/", mustWork = TRUE)
  inferred_name <- if (is.null(block_name)) basename(normalized_path) else block_name
  parse_fm_numeric_text(readChar(normalized_path, nchars = file.info(normalized_path)$size, useBytes = TRUE), block_name = inferred_name)
}

parse_fmexog_text <- function(text) {
  lines <- strsplit(gsub("\r", "", text), "\n", fixed = TRUE)[[1]]
  rows <- list()

  current_window_start <- NULL
  current_window_end <- NULL
  current_window_size <- NULL
  in_changevar <- FALSE

  pending_scalar <- FALSE
  pending_variable <- NULL
  pending_method <- NULL
  pending_values <- numeric()
  pending_vector <- FALSE

  clean_token <- function(token) {
    raw <- trimws(token)
    raw <- gsub("^['\"]|['\"]$", "", raw)
    sub(";\\s*$", "", raw)
  }
  extract_numbers <- function(line) {
    matches <- gregexpr(
      "[+-]?(?:\\d+(?:\\.\\d*)?|\\.\\d+)(?:[Ee][+-]?\\d+)?",
      line,
      perl = TRUE
    )
    found <- regmatches(line, matches)[[1]]
    if (!length(found)) {
      return(numeric())
    }
    as.numeric(found)
  }
  is_numeric_token <- function(token) {
    grepl(
      "^[+-]?(?:\\d+(?:\\.\\d*)?|\\.\\d+)(?:[Ee][+-]?\\d+)?$",
      clean_token(token),
      perl = TRUE
    )
  }
  close_pending <- function() {
    if (is.null(pending_variable)) {
      return(invisible(NULL))
    }
    if (is.null(current_window_start) || is.null(current_window_end)) {
      stopf("CHANGEVAR instruction found outside an SMPL window")
    }
    if (!length(pending_values)) {
      stopf("No values parsed for instruction %s", pending_variable)
    }
    rows[[length(rows) + 1L]] <<- list(
      window_start = current_window_start,
      window_end = current_window_end,
      variable = pending_variable,
      method = pending_method %||% "",
      is_vector = isTRUE(pending_vector),
      n_values = length(pending_values),
      values = as.numeric(pending_values)
    )
    pending_scalar <<- FALSE
    pending_variable <<- NULL
    pending_method <<- NULL
    pending_values <<- numeric()
    pending_vector <<- FALSE
    invisible(NULL)
  }

  for (raw_line in lines) {
    line <- trimws(raw_line)
    if (!nzchar(line) || grepl("^\\s*@", line)) {
      next
    }
    line_has_semicolon <- grepl(";", line, fixed = TRUE)
    tokens <- strsplit(line, "\\s+", perl = TRUE)[[1]]
    tokens <- tokens[nzchar(tokens)]
    if (!length(tokens)) {
      next
    }

    first <- toupper(clean_token(tokens[[1]]))
    if (identical(first, "SMPL")) {
      if (length(tokens) < 3L) {
        stopf("SMPL statements require start and end periods")
      }
      close_pending()
      current_window_start <- clean_token(tokens[[2]])
      current_window_end <- clean_token(tokens[[3]])
      current_window_size <- length(generate_smpl_periods(current_window_start, current_window_end))
      in_changevar <- FALSE
      next
    }
    if (identical(first, "CHANGEVAR")) {
      if (is.null(current_window_start)) {
        stopf("CHANGEVAR encountered before SMPL statement")
      }
      close_pending()
      in_changevar <- TRUE
      next
    }
    if (identical(first, "RETURN")) {
      close_pending()
      break
    }
    if (!in_changevar) {
      next
    }

    if (pending_scalar || pending_vector) {
      if (!is_numeric_token(tokens[[1]])) {
        close_pending()
      } else {
        values <- extract_numbers(line)
        if (length(values)) {
          pending_values <- c(pending_values, values)
          if (pending_scalar && length(pending_values) > 1L) {
            pending_scalar <- FALSE
            pending_vector <- TRUE
          }
        }
        if (isTRUE(pending_vector) &&
          !is.null(current_window_size) &&
          length(pending_values) >= current_window_size) {
          if (length(pending_values) > current_window_size) {
            stopf(
              "Too many values parsed for vector instruction %s: expected %d, found %d",
              pending_variable,
              current_window_size,
              length(pending_values)
            )
          }
          close_pending()
          next
        }
        if (line_has_semicolon) {
          close_pending()
        }
        next
      }
    }

    variable <- clean_token(tokens[[1]])
    if (!nzchar(variable)) {
      next
    }
    method <- NULL
    is_vector <- FALSE
    value_text <- line

    if (length(tokens) >= 2L && !is_numeric_token(tokens[[2]])) {
      method <- clean_token(tokens[[2]])
      split_idx <- regexpr(tokens[[2]], line, fixed = TRUE)[[1]]
      value_text <- substr(line, split_idx + nchar(tokens[[2]]), nchar(line))
      if (line_has_semicolon) {
        is_vector <- TRUE
      }
    }
    if (!line_has_semicolon && length(tokens) == 1L) {
      is_vector <- FALSE
    } else if (line_has_semicolon && length(tokens) == 1L) {
      is_vector <- TRUE
    }

    extracted <- extract_numbers(value_text)
    if (isTRUE(is_vector)) {
      pending_scalar <- FALSE
      pending_vector <- TRUE
      pending_variable <- variable
      pending_method <- method
      pending_values <- extracted
      if (length(pending_values) && line_has_semicolon) {
        close_pending()
      }
      next
    }

    pending_scalar <- TRUE
    pending_vector <- FALSE
    pending_variable <- variable
    pending_method <- method
    pending_values <- extracted
    if (length(pending_values)) {
      close_pending()
    }
  }

  close_pending()
  if (!length(rows)) {
    empty <- data.frame(
      window_start = character(),
      window_end = character(),
      variable = character(),
      method = character(),
      is_vector = logical(),
      n_values = integer(),
      stringsAsFactors = FALSE
    )
    empty$values <- I(list())
    return(empty)
  }
  out <- data.frame(
    window_start = vapply(rows, `[[`, "", "window_start"),
    window_end = vapply(rows, `[[`, "", "window_end"),
    variable = vapply(rows, `[[`, "", "variable"),
    method = vapply(rows, `[[`, "", "method"),
    is_vector = vapply(rows, `[[`, FALSE, "is_vector"),
    n_values = vapply(rows, `[[`, integer(1), "n_values"),
    stringsAsFactors = FALSE
  )
  out$values <- I(lapply(rows, `[[`, "values"))
  out
}

parse_fmexog_file <- function(path) {
  normalized_path <- normalizePath(path, winslash = "/", mustWork = TRUE)
  parse_fmexog_text(readChar(normalized_path, nchars = file.info(normalized_path)$size, useBytes = TRUE))
}

merge_fm_numeric_frames <- function(base_frame, loaded_frame) {
  if (!nrow(loaded_frame)) {
    return(base_frame)
  }
  if (!nrow(base_frame)) {
    return(loaded_frame)
  }

  all_periods <- sort(unique(c(as.character(base_frame$period), as.character(loaded_frame$period))))
  merged <- data.frame(
    period = all_periods,
    stringsAsFactors = FALSE,
    check.names = FALSE
  )

  merge_one <- function(frame, name, carry_forward = FALSE) {
    values <- rep(NA_real_, length(all_periods))
    names(values) <- all_periods
    if (name %in% names(frame)) {
      values[as.character(frame$period)] <- as.numeric(frame[[name]])
    }
    if (carry_forward) {
      last_value <- NA_real_
      for (idx in seq_along(values)) {
        if (is.finite(values[[idx]])) {
          last_value <- values[[idx]]
        } else if (is.finite(last_value)) {
          values[[idx]] <- last_value
        }
      }
    }
    as.numeric(values)
  }

  base_columns <- setdiff(names(base_frame), "period")
  loaded_columns <- setdiff(names(loaded_frame), "period")
  for (name in base_columns) {
    merged[[name]] <- merge_one(base_frame, name, carry_forward = TRUE)
  }
  for (name in loaded_columns) {
    if (!(name %in% names(merged))) {
      merged[[name]] <- 0
    }
    incoming <- merge_one(loaded_frame, name, carry_forward = FALSE)
    base_missing <- is.na(merged[[name]])
    merged[[name]][base_missing] <- 0
    have_incoming <- !is.na(incoming)
    merged[[name]][have_incoming] <- incoming[have_incoming]
  }
  merged
}

apply_fmexog_rows <- function(base_frame, rows) {
  if (!nrow(rows)) {
    return(base_frame)
  }
  working <- base_frame
  working$period <- as.character(working$period)
  index_positions <- seq_along(working$period)
  names(index_positions) <- working$period

  first_finite <- function(values) {
    for (value in values) {
      if (is.finite(value)) {
        return(as.numeric(value))
      }
    }
    0
  }

  for (row_idx in seq_len(nrow(rows))) {
    row <- rows[row_idx, , drop = FALSE]
    variable <- trimws(as.character(row$variable[[1]]))
    if (!nzchar(variable)) {
      next
    }
    method <- toupper(trimws(as.character(row$method[[1]])))
    if (!nzchar(method)) {
      method <- "CHGSAMEPCT"
    }
    values <- as.numeric(unlist(row$values[[1]]))
    if (!length(values)) {
      next
    }

    periods <- generate_smpl_periods(row$window_start[[1]], row$window_end[[1]])
    periods <- periods[periods %in% working$period]
    if (!length(periods)) {
      next
    }

    if (!(variable %in% names(working))) {
      working[[variable]] <- NA_real_
    }
    is_vector <- isTRUE(row$is_vector[[1]])

    if (is_vector) {
      if (length(values) == length(periods)) {
        padded <- values
      } else if (length(values) > 1L) {
        padded <- values[seq_len(min(length(values), length(periods)))]
        if (length(padded) < length(periods)) {
          padded <- c(padded, rep(tail(padded, 1L), length(periods) - length(padded)))
        }
      } else {
        padded <- rep(values[[1]], length(periods))
      }

      if (method %in% c("ADDDIFABS", "ADDDIFPCT")) {
        existing <- as.numeric(working[match(periods, working$period), variable])
        existing[!is.finite(existing)] <- 0
        updated <- if (identical(method, "ADDDIFABS")) {
          existing + padded
        } else {
          existing * (1 + padded)
        }
        working[match(periods, working$period), variable] <- updated
        next
      }

      working[match(periods, working$period), variable] <- padded
      next
    }

    scalar <- as.numeric(values[[1]])
    if (identical(method, "SAMEVALUE")) {
      working[match(periods, working$period), variable] <- scalar
      next
    }

    if (method %in% c("CHGSAMEABS", "CHGSAMEPCT")) {
      series <- as.numeric(working[[variable]])
      names(series) <- working$period
      for (period in periods) {
        position <- index_positions[[period]]
        prev_value <- NA_real_
        if (!is.null(position) && position > 1L) {
          prev_period <- working$period[[position - 1L]]
          prev_value <- as.numeric(series[[prev_period]])
        }
        current_value <- if (period %in% names(series)) as.numeric(series[[period]]) else NA_real_
        baseline <- first_finite(c(prev_value, current_value, 0))
        updated <- if (identical(method, "CHGSAMEABS")) baseline + scalar else baseline * (1 + scalar)
        working[working$period == period, variable] <- updated
        series[[period]] <- updated
      }
      next
    }

    if (method %in% c("ADDDIFABS", "ADDDIFPCT")) {
      series <- as.numeric(working[[variable]])
      names(series) <- working$period
      for (period in periods) {
        current_value <- if (period %in% names(series)) as.numeric(series[[period]]) else NA_real_
        baseline <- first_finite(c(current_value, 0))
        updated <- if (identical(method, "ADDDIFABS")) baseline + scalar else baseline * (1 + scalar)
        working[working$period == period, variable] <- updated
        series[[period]] <- updated
      }
      next
    }

    working[match(periods, working$period), variable] <- scalar
  }

  working
}
